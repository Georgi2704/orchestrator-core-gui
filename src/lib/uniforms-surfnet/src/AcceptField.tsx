/*
 * Copyright 2019-2020 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import "./AcceptField.scss";

import CheckBox from "components/CheckBox";
import I18n, { TranslateOptions } from "i18n-js";
import React, { HTMLProps, useReducer } from "react";
import { Override, connectField } from "uniforms";

type AcceptItem = (string | TranslateOptions)[];
type AcceptValue = "SKIPPED" | "ACCEPTED" | "INCOMPLETE";

export type AcceptFieldProps = Override<
    HTMLProps<HTMLDivElement>,
    {
        disabled: boolean;
        id: string;
        name: string;
        onChange(value?: AcceptValue): void;
        value?: AcceptValue;
        error?: boolean;
        showInlineError?: boolean;
        errorMessage?: string;
        data?: AcceptItem[];
    }
>;

interface AcceptState {
    checks: { [index: number]: boolean };
    skip: boolean;
    allChecked: boolean;
}

interface Action {
    field: number;
    type: string;
    value: boolean;
}

function Accept({
    disabled,
    id,
    name,
    onChange,
    value,
    error,
    showInlineError,
    errorMessage,
    data,
    ...props
}: AcceptFieldProps) {
    const legacy = !data;
    const i18nBaseKey = data ? `forms.fields.${name}_accept` : "forms.fields";

    data = data ?? [
        [name, "label", {}],
        [`${name}_info`, "info", {}],
        [name, "checkbox", {}]
    ];

    let [state, dispatch] = useReducer(
        (state: AcceptState, action: Action) => {
            if (action.type === "skip") {
                state.skip = action.value;
            } else {
                state.checks[action.field] = action.value;
            }

            // We intentionally skip optional checkboxes here
            state.allChecked = data!
                .map((entry: AcceptItem, index: number) => [entry, state.checks[index] || false])
                .filter((entry: (boolean | AcceptItem)[]) => (entry[0] as AcceptItem)[1].endsWith("checkbox"))
                .map((entry: (boolean | AcceptItem)[]) => entry[1] as boolean)
                .every((check: boolean) => check);

            onChange(state.skip ? "SKIPPED" : state.allChecked ? "ACCEPTED" : "INCOMPLETE");

            return { ...state };
        },
        { checks: {}, skip: false, allChecked: false }
    );

    return (
        <section id={id} className="accept-field">
            {data.map((entry: any[], index: number) => {
                const label = I18n.t(`${i18nBaseKey}.${entry[0]}`, entry[2] as TranslateOptions);

                switch (entry[1]) {
                    case "label":
                        return (
                            <div key={index}>
                                <label>{label}</label>
                            </div>
                        );
                    case "info":
                        return (
                            <div key={index}>
                                <em>{label}</em>
                            </div>
                        );
                    case "url":
                        return (
                            <div key={index}>
                                <a href={entry[0] as string} target="_blank" rel="noopener noreferrer">
                                    {entry[0]}
                                </a>
                            </div>
                        );
                    case "warning":
                        return (
                            <div key={index}>
                                <label className="warning">{label}</label>
                            </div>
                        );
                    case "skip":
                        return (
                            <CheckBox
                                key={index}
                                name={entry[0]}
                                onChange={(e: React.FormEvent<HTMLInputElement>) => {
                                    const target = e.target as HTMLInputElement;
                                    dispatch({ field: index, type: "skip", value: target.checked });
                                }}
                                value={state.skip}
                                info={label}
                                className={"skip"}
                            />
                        );
                    default:
                        return (
                            <CheckBox
                                key={index}
                                name={entry[0] + (legacy ? "" : index)} // Index needed to allow checkboxes with same name (we can skip this in legacy mode)
                                className={entry[1].startsWith(">") ? "level_2" : undefined}
                                onChange={(e: React.FormEvent<HTMLInputElement>) => {
                                    const target = e.target as HTMLInputElement;

                                    dispatch({ field: index, type: "check", value: target.checked });
                                }}
                                value={state.checks[index]}
                                info={label}
                                disabled={state.skip || disabled}
                            />
                        );
                }
            })}

            {error && showInlineError && (
                <em className="error">
                    <div className="backend-validation">{errorMessage}</div>
                </em>
            )}
        </section>
    );
}

export default connectField(Accept);
