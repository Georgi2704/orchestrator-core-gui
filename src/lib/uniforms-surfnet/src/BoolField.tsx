/*
 * Copyright 2019-2020 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import "./BoolField.scss";

import React, { HTMLProps, Ref } from "react";
import { Override, connectField, filterDOMProps } from "uniforms";

export type BoolFieldProps = Override<
    HTMLProps<HTMLDivElement>,
    {
        disabled: boolean;
        id: string;
        inputRef?: Ref<HTMLInputElement>;
        label: string;
        description?: string;
        name: string;
        onChange(value?: boolean): void;
        value?: boolean;
        error?: boolean;
        showInlineError?: boolean;
        errorMessage?: string;
    }
>;

function Bool({
    disabled,
    id,
    inputRef,
    label,
    description,
    name,
    onChange,
    value,
    error,
    showInlineError,
    errorMessage,
    ...props
}: BoolFieldProps) {
    return (
        <section {...filterDOMProps(props)} className="bool-field">
            <input
                checked={value || false}
                disabled={disabled}
                id={id}
                name={name}
                onChange={
                    disabled
                        ? undefined
                        : () => {
                              onChange(!value);
                          }
                }
                ref={inputRef}
                type="checkbox"
            />
            <label htmlFor={id}>
                <span tabIndex={0}>
                    <i className="fa fa-check" />
                </span>
            </label>
            {label && (
                <label className="info" htmlFor={id}>
                    {label}
                </label>
            )}
            {error && showInlineError && (
                <em className="error">
                    <div className="backend-validation">{errorMessage}</div>
                </em>
            )}
        </section>
    );
}

export default connectField(Bool);
