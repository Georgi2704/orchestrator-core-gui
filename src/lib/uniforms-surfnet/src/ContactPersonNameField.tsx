/*
 * Copyright 2019-2020 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import { contacts } from "api";
import Autocomplete from "components/Autocomplete";
import I18n from "i18n-js";
import { isFunction } from "lodash";
import get from "lodash/get";
import React, { Ref, useEffect, useState } from "react";
import { connectField, filterDOMProps, joinName, useForm } from "uniforms";
import { ContactPerson } from "utils/types";

import { isEmpty, stop } from "../../../utils/Utils";
import { FieldProps } from "./types";

export type ContactPersonNameFieldProps = FieldProps<string, { organisationId?: string; organisationKey?: string }>;

function ContactPersonName({
    disabled,
    id,
    inputRef = React.createRef() as Ref<HTMLInputElement>,
    label,
    description,
    name,
    onChange,
    placeholder,
    value,
    error,
    showInlineError,
    errorMessage,
    organisationId,
    organisationKey,
    ...props
}: ContactPersonNameFieldProps) {
    const { model, onChange: formOnChange } = useForm();

    const organisationIdValue = organisationId || get(model, organisationKey || "organisation");
    const contactsPersonFieldNameArray = joinName(null, name).slice(0, -1);
    const emailFieldName = joinName(contactsPersonFieldNameArray, "email");
    const phoneFieldName = joinName(contactsPersonFieldNameArray, "phone");
    const contactsFieldName = joinName(contactsPersonFieldNameArray.slice(0, -1));

    let [displayAutocomplete, setDisplayAutocomplete] = useState({});
    let [contactPersons, setContactPersons] = useState<ContactPerson[]>([]);
    let [suggestions, setSuggestions] = useState<ContactPerson[]>([]);
    let [selectedIndex, setSelectedIndex] = useState(-1);

    useEffect(() => {
        if (organisationIdValue) {
            contacts(organisationIdValue).then(setContactPersons);
        }
    }, [organisationIdValue]);

    useEffect(() => {
        // Set focus to the last name component to be created
        if (!isFunction(inputRef)) {
            inputRef!.current?.focus();
        }
    }, []); // eslint-disable-line react-hooks/exhaustive-deps

    function onChangeInternal(e: React.FormEvent<HTMLInputElement>) {
        stop(e);
        const target = e.target as HTMLInputElement;
        const value = target.value;

        onChange(value);

        const persons: ContactPerson[] = get(model, contactsFieldName);

        const filteredSuggestions = !value
            ? []
            : contactPersons
                  .filter(item => item.name.toLowerCase().indexOf(value.toLowerCase()) > -1)
                  .filter(item => !persons.some(person => person.email === item.email));
        setDisplayAutocomplete(!isEmpty(filteredSuggestions));
        setSuggestions(filteredSuggestions);
    }

    function itemSelected(item: ContactPerson, personIndex: number) {
        onChange(item.name || "");
        formOnChange(emailFieldName, item.email || "");
        formOnChange(phoneFieldName, item.phone || "");
        setDisplayAutocomplete(false);
        setSelectedIndex(-1);
    }

    function onAutocompleteKeyDown(e: React.KeyboardEvent<HTMLElement>) {
        if (!suggestions) {
            return;
        }
        if (e.keyCode === 40 && selectedIndex < suggestions.length - 1) {
            //keyDown
            stop(e);
            setSelectedIndex(selectedIndex++);
        }
        if (e.keyCode === 38 && selectedIndex >= 0) {
            //keyUp
            stop(e);
            setSelectedIndex(selectedIndex--);
        }
        if (e.keyCode === 13) {
            //enter
            if (selectedIndex >= 0) {
                stop(e);
                itemSelected(suggestions[selectedIndex], 0);
            }
        }
        if (e.keyCode === 27) {
            //escape
            stop(e);
            setDisplayAutocomplete(false);
            setSelectedIndex(-1);
        }
    }

    function onBlurAutoComplete(e: React.FocusEvent<HTMLElement>) {
        stop(e);
        setTimeout(() => setDisplayAutocomplete(false), 350);
    }

    return (
        <section {...filterDOMProps(props)}>
            {label && (
                <label htmlFor={id}>
                    {label}
                    {description && <em>{description}</em>}
                </label>
            )}
            <div className="autocomplete-container">
                <input
                    disabled={disabled}
                    id={id}
                    name={name}
                    onChange={onChangeInternal}
                    placeholder={placeholder || I18n.t("forms.widgets.contactPersonName.placeholder")}
                    ref={inputRef}
                    type="text"
                    value={value ?? ""}
                    onKeyDown={onAutocompleteKeyDown}
                    onBlur={onBlurAutoComplete}
                ></input>
                {displayAutocomplete && (
                    <Autocomplete
                        query={value ?? ""}
                        itemSelected={itemSelected}
                        selectedItem={selectedIndex}
                        suggestions={suggestions}
                        personIndex={0}
                    />
                )}{" "}
            </div>
            {error && showInlineError && (
                <em className="error">
                    <div className="backend-validation">{errorMessage}</div>
                </em>
            )}
        </section>
    );
}

export default connectField(ContactPersonName, { kind: "leaf" });
