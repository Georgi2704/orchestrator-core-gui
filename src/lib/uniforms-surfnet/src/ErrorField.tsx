import React, { HTMLProps, ReactNode } from "react";
import { connectField, filterDOMProps } from "uniforms";

import { Override } from "./utils";

export type ErrorFieldProps = Override<
    HTMLProps<HTMLDivElement>,
    {
        children?: ReactNode;
        error?: any;
        errorMessage?: string;
    }
>;

function Error({ children, error, errorMessage, ...props }: ErrorFieldProps) {
    return !error ? null : <div {...filterDOMProps(props)}>{children || errorMessage}</div>;
}

export default connectField(Error, { initialValue: false });
