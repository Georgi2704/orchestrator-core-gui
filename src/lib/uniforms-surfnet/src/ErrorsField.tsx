import React from "react";
import { BaseField, filterDOMProps, nothing } from "uniforms";

const ErrorsField = ({ children, ...props }: any, { uniforms: { error, schema } }: any) =>
    !error && !children ? (
        nothing
    ) : (
        <div {...filterDOMProps(props)}>
            {children}

            <ul>
                {schema.getErrorMessages(error).map((message: string, index: number) => (
                    <li key={index}>{message}</li>
                ))}
            </ul>
        </div>
    );

ErrorsField.contextTypes = BaseField.contextTypes;

export default ErrorsField;
