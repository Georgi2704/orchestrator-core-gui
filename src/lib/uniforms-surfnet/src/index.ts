//@ts-nocheck
export { default as AutoField } from "./AutoField";
export { default as AutoFields } from "./AutoFields";
export { default as BoolField } from "./BoolField";
export { default as DateField } from "./DateField";
export { default as ErrorField } from "./ErrorField";
export { default as ErrorsField } from "./ErrorsField";
export { default as HiddenField } from "./HiddenField";
export { default as ListAddField } from "./ListAddField";
export { default as ListDelField } from "./ListDelField";
export { default as ListField } from "./ListField";
export { default as ListItemField } from "./ListItemField";
export { default as LongTextField } from "./LongTextField";
export { default as NestField } from "./NestField";
export { default as NumField } from "./NumField";
export { default as RadioField } from "./RadioField";
export { default as SelectField } from "./SelectField";
export { default as SubmitField } from "./SubmitField";
export { default as TextField } from "./TextField";

export { default as ContactPersonNameField } from "./ContactPersonNameField";
export { default as ImsPortIdField } from "./ImsPortIdField";
export { default as VlanField } from "./VlanField";

export { default as LocationCodeField } from "./logic/LocationCodeField";
export { default as OrganisationField } from "./logic/OrganisationField";
export { default as ProductField } from "./logic/ProductField";
export { default as SubscriptionField } from "./logic/SubscriptionField";
