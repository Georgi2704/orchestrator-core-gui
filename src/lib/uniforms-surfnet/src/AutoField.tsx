/*
 * Copyright 2019-2020 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import invariant from "invariant";
import { ComponentType, createElement } from "react";
import { Override, connectField, useField } from "uniforms";

import BoolField from "./BoolField";
import DateField from "./DateField";
import ListField from "./ListField";
import NestField from "./NestField";
import NumField from "./NumField";
import RadioField from "./RadioField";
import SelectField from "./SelectField";
import TextField from "./TextField";

export type AutoFieldProps = Override<
    Record<string, unknown>,
    {
        component?: ComponentType<any> | ReturnType<typeof connectField>;
        name: string;
    }
>;

export default function AutoField(originalProps: AutoFieldProps) {
    const props = useField(originalProps.name, originalProps)[0];
    const { allowedValues, checkboxes, fieldType } = props;
    let { component } = props;

    if (component === undefined) {
        if (allowedValues) {
            if (checkboxes && fieldType !== Array) {
                component = RadioField;
            } else {
                component = SelectField;
            }
        } else {
            switch (fieldType) {
                case Array:
                    component = ListField;
                    break;
                case Boolean:
                    component = BoolField;
                    break;
                case Date:
                    component = DateField;
                    break;
                case Number:
                    component = NumField;
                    break;
                case Object:
                    component = NestField;
                    break;
                case String:
                    component = TextField;
                    break;
            }

            invariant(component, "Unsupported field type: %s", fieldType);
        }
    }

    return "options" in component && component.options?.kind === "leaf"
        ? createElement(component.Component, props)
        : createElement(component, originalProps);
}
