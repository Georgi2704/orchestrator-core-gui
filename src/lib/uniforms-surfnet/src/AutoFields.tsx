import { ComponentType, createElement } from "react";
import { useForm } from "uniforms";

import AutoField from "./AutoField";

export type AutoFieldsProps = {
    autoField?: ComponentType<{ name: string; className: string; id: string }>;
    element?: ComponentType<{ className: string }> | string;
    fields?: string[];
    omitFields?: string[];
};

export default function AutoFields({ autoField, element, fields, omitFields, ...props }: AutoFieldsProps) {
    const { schema } = useForm();

    return createElement(
        element!,
        { className: "form-step", ...props },
        (fields ?? schema.getSubfields())
            .filter(field => !omitFields!.includes(field))
            .map(field =>
                createElement(autoField!, { key: field, name: field, className: field, id: `input-${field}` })
            )
    );
}

AutoFields.defaultProps = {
    autoField: AutoField,
    element: "section",
    omitFields: []
};
