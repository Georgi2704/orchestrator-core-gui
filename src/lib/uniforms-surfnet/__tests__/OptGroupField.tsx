/*
 * Copyright 2019-2020 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import React from "react";

import { OptGroupField } from "../src";
import createContext from "./_createContext";
import mount from "./_mount";

test("<OptGroupField> - renders an the correct fields when disabled", () => {
    const element = <OptGroupField name="x" />;
    const wrapper = mount(
        element,
        createContext({
            x: { type: Object },
            "x.enabled": { type: Boolean },
            "x.b": { type: Number },
            "x.c": { type: Number }
        })
    );

    expect(wrapper.find("input")).toHaveLength(1);
    expect(
        wrapper
            .find("input")
            .at(0)
            .prop("name")
    ).toBe("x.enabled");
});

test("<OptGroupField> - renders an the correct fields when enabled", () => {
    const element = <OptGroupField name="x" />;
    const wrapper = mount(
        element,
        createContext(
            {
                x: { type: Object },
                "x.enabled": { type: Boolean },
                "x.b": { type: Number },
                "x.c": { type: Number }
            },
            { model: { x: { enabled: true } } }
        )
    );

    expect(wrapper.find("input")).toHaveLength(3);
    expect(
        wrapper
            .find("input")
            .at(0)
            .prop("name")
    ).toBe("x.enabled");
    expect(
        wrapper
            .find("input")
            .at(1)
            .prop("name")
    ).toBe("x.b");
    expect(
        wrapper
            .find("input")
            .at(2)
            .prop("name")
    ).toBe("x.c");
});

test("<OptGroupField> - renders a label", () => {
    const element = <OptGroupField name="x" label="y" />;
    const wrapper = mount(
        element,
        createContext({
            x: { type: Object },
            "x.enabled": { type: Boolean },
            "x.b": { type: Number }
        })
    );

    expect(wrapper.find("label")).toHaveLength(3);
    expect(
        wrapper
            .find("label")
            .at(0)
            .text()
    ).toBe('[missing "en.forms.fields.x.title" translation][missing "en.forms.fields.x.info" translation]');
});

test("<OptGroupField> - renders a wrapper with unknown props", () => {
    const element = <OptGroupField name="x" data-x="x" data-y="y" data-z="z" />;
    const wrapper = mount(
        element,
        createContext({
            x: { type: Object },
            "x.enabled": { type: Boolean },
            "x.b": { type: Number }
        })
    );

    expect(
        wrapper
            .find("section")
            .at(0)
            .prop("data-x")
    ).toBe("x");
    expect(
        wrapper
            .find("section")
            .at(0)
            .prop("data-y")
    ).toBe("y");
    expect(
        wrapper
            .find("section")
            .at(0)
            .prop("data-z")
    ).toBe("z");
});
