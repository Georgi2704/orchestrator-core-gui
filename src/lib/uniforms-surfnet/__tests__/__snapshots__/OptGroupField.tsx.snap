// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OptGroupField> - renders an the correct fields when disabled 1`] = `
"<OptGroupField name=\\"x\\">
  <OptGroup name=\\"x\\" id=\\"uniforms-0000-0000\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Object], optional: false, label: 'X' }} fieldType={[Function: Object]} fields={[ 'enabled', 'b', 'c' ]} onChange={[Function]} value={{}} required={true}>
    <section id=\\"uniforms-0000-0000\\" required={true} className=\\" optgroup-field\\">
      <label>
        [missing &quot;en.forms.fields.x.title&quot; translation]
        <em>
          [missing &quot;en.forms.fields.x.info&quot; translation]
        </em>
      </label>
      <BoolField name=\\"enabled\\">
        <Bool name=\\"x.enabled\\" id=\\"uniforms-0000-0002\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Boolean], optional: false, label: 'Enabled' }} fieldType={[Function: Boolean]} fields={[]} onChange={[Function]} value={[undefined]} required={true}>
          <section required={true} className=\\" bool-field\\">
            <input checked={false} disabled={false} id=\\"uniforms-0000-0002\\" name=\\"x.enabled\\" onChange={[Function]} type=\\"checkbox\\" />
            <label htmlFor=\\"uniforms-0000-0002\\">
              <span tabIndex={0}>
                <i className=\\"fa fa-check\\" />
              </span>
            </label>
          </section>
        </Bool>
      </BoolField>
    </section>
  </OptGroup>
</OptGroupField>"
`;

exports[`<OptGroupField> - renders an the correct fields when enabled 1`] = `
"<OptGroupField name=\\"x\\">
  <OptGroup name=\\"x\\" id=\\"uniforms-0000-0003\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Object], optional: false, label: 'X' }} fieldType={[Function: Object]} fields={[ 'enabled', 'b', 'c' ]} onChange={[Function]} value={{ enabled: true }} required={true}>
    <section id=\\"uniforms-0000-0003\\" required={true} className=\\" optgroup-field\\">
      <label>
        [missing &quot;en.forms.fields.x.title&quot; translation]
        <em>
          [missing &quot;en.forms.fields.x.info&quot; translation]
        </em>
      </label>
      <BoolField name=\\"enabled\\">
        <Bool name=\\"x.enabled\\" id=\\"uniforms-0000-0005\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Boolean], optional: false, label: 'Enabled' }} fieldType={[Function: Boolean]} fields={[]} onChange={[Function]} value={true} required={true}>
          <section required={true} className=\\" bool-field\\">
            <input checked={true} disabled={false} id=\\"uniforms-0000-0005\\" name=\\"x.enabled\\" onChange={[Function]} type=\\"checkbox\\" />
            <label htmlFor=\\"uniforms-0000-0005\\">
              <span tabIndex={0}>
                <i className=\\"fa fa-check\\" />
              </span>
            </label>
          </section>
        </Bool>
      </BoolField>
      <AutoField name=\\"b\\">
        <Num id=\\"uniforms-0000-0006\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Number], optional: false, label: 'B' }} fieldType={[Function: Number]} fields={[]} onChange={[Function]} value={[undefined]} required={true} decimal={true} name=\\"x.b\\">
          <div required={true}>
            <NumericInput id=\\"uniforms-0000-0006\\" name=\\"x.b\\" placeholder=\\"\\" onChange={[Function: onChange]} min={-9007199254740991} max={9007199254740991} step={1} precision={0} value=\\"\\" strict={false} disabled={false} parse={null} format={null} mobile=\\"auto\\" componentClass=\\"input\\" style={{}}>
              <span style={{ position: 'relative', display: 'inline-block' }} className=\\"react-numeric-input\\" onMouseUp={[Function: bound stop]} onMouseLeave={[Function: bound stop]}>
                <input type=\\"text\\" style={{ paddingRight: '3ex', boxSizing: 'border-box', fontSize: 'inherit', border: '1px solid #ccc', borderRadius: 2, paddingLeft: 4, display: 'block', WebkitAppearance: 'none', lineHeight: 'normal' }} id=\\"uniforms-0000-0006\\" name=\\"x.b\\" placeholder=\\"\\" onChange={[Function: onChange]} disabled={false} value=\\"\\" onKeyDown={[Function: bound _onKeyDown]} onInput={[Function: onInput]} onSelect={[Function: onSelect]} onFocus={[Function: onFocus]} onBlur={[Function: onBlur]} />
                <b onMouseEnter={[Function: onMouseEnter]} onMouseDown={[Function: onMouseDown]} onMouseUp={[Function: onMouseUp]} onMouseLeave={[Function: onMouseLeave]} onTouchStart={[Function: bound onTouchStart]} onTouchEnd={[Function: bound onTouchEnd]} style={{ position: 'absolute', right: 2, width: '2.26ex', borderColor: 'rgba(0,0,0,.1)', borderStyle: 'solid', textAlign: 'center', cursor: 'default', transition: 'all 0.1s', background: 'rgba(0,0,0,.1)', boxShadow: '-1px -1px 3px rgba(0,0,0,.1) inset,1px 1px 3px rgba(255,255,255,.7) inset', top: 2, bottom: '50%', borderRadius: '2px 2px 0 0', borderWidth: '1px 1px 0 1px' }}>
                  <i style={{ position: 'absolute', top: '50%', left: '50%', width: 0, height: 0, borderWidth: '0 0.6ex 0.6ex 0.6ex', borderColor: 'transparent transparent rgba(0, 0, 0, 0.7)', borderStyle: 'solid', margin: '-0.3ex 0 0 -0.56ex' }} />
                </b>
                <b onMouseEnter={[Function: onMouseEnter]} onMouseDown={[Function: onMouseDown]} onMouseUp={[Function: onMouseUp]} onMouseLeave={[Function: onMouseLeave]} onTouchStart={[Function: bound onTouchStart]} onTouchEnd={[Function: bound onTouchEnd]} style={{ position: 'absolute', right: 2, width: '2.26ex', borderColor: 'rgba(0,0,0,.1)', borderStyle: 'solid', textAlign: 'center', cursor: 'default', transition: 'all 0.1s', background: 'rgba(0,0,0,.1)', boxShadow: '-1px -1px 3px rgba(0,0,0,.1) inset,1px 1px 3px rgba(255,255,255,.7) inset', top: '50%', bottom: 2, borderRadius: '0 0 2px 2px', borderWidth: '0 1px 1px 1px' }}>
                  <i style={{ position: 'absolute', top: '50%', left: '50%', width: 0, height: 0, borderWidth: '0.6ex 0.6ex 0 0.6ex', borderColor: 'rgba(0, 0, 0, 0.7) transparent transparent', borderStyle: 'solid', margin: '-0.3ex 0 0 -0.56ex' }} />
                </b>
              </span>
            </NumericInput>
          </div>
        </Num>
      </AutoField>
      <AutoField name=\\"c\\">
        <Num id=\\"uniforms-0000-0007\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Number], optional: false, label: 'C' }} fieldType={[Function: Number]} fields={[]} onChange={[Function]} value={[undefined]} required={true} decimal={true} name=\\"x.c\\">
          <div required={true}>
            <NumericInput id=\\"uniforms-0000-0007\\" name=\\"x.c\\" placeholder=\\"\\" onChange={[Function: onChange]} min={-9007199254740991} max={9007199254740991} step={1} precision={0} value=\\"\\" strict={false} disabled={false} parse={null} format={null} mobile=\\"auto\\" componentClass=\\"input\\" style={{}}>
              <span style={{ position: 'relative', display: 'inline-block' }} className=\\"react-numeric-input\\" onMouseUp={[Function: bound stop]} onMouseLeave={[Function: bound stop]}>
                <input type=\\"text\\" style={{ paddingRight: '3ex', boxSizing: 'border-box', fontSize: 'inherit', border: '1px solid #ccc', borderRadius: 2, paddingLeft: 4, display: 'block', WebkitAppearance: 'none', lineHeight: 'normal' }} id=\\"uniforms-0000-0007\\" name=\\"x.c\\" placeholder=\\"\\" onChange={[Function: onChange]} disabled={false} value=\\"\\" onKeyDown={[Function: bound _onKeyDown]} onInput={[Function: onInput]} onSelect={[Function: onSelect]} onFocus={[Function: onFocus]} onBlur={[Function: onBlur]} />
                <b onMouseEnter={[Function: onMouseEnter]} onMouseDown={[Function: onMouseDown]} onMouseUp={[Function: onMouseUp]} onMouseLeave={[Function: onMouseLeave]} onTouchStart={[Function: bound onTouchStart]} onTouchEnd={[Function: bound onTouchEnd]} style={{ position: 'absolute', right: 2, width: '2.26ex', borderColor: 'rgba(0,0,0,.1)', borderStyle: 'solid', textAlign: 'center', cursor: 'default', transition: 'all 0.1s', background: 'rgba(0,0,0,.1)', boxShadow: '-1px -1px 3px rgba(0,0,0,.1) inset,1px 1px 3px rgba(255,255,255,.7) inset', top: 2, bottom: '50%', borderRadius: '2px 2px 0 0', borderWidth: '1px 1px 0 1px' }}>
                  <i style={{ position: 'absolute', top: '50%', left: '50%', width: 0, height: 0, borderWidth: '0 0.6ex 0.6ex 0.6ex', borderColor: 'transparent transparent rgba(0, 0, 0, 0.7)', borderStyle: 'solid', margin: '-0.3ex 0 0 -0.56ex' }} />
                </b>
                <b onMouseEnter={[Function: onMouseEnter]} onMouseDown={[Function: onMouseDown]} onMouseUp={[Function: onMouseUp]} onMouseLeave={[Function: onMouseLeave]} onTouchStart={[Function: bound onTouchStart]} onTouchEnd={[Function: bound onTouchEnd]} style={{ position: 'absolute', right: 2, width: '2.26ex', borderColor: 'rgba(0,0,0,.1)', borderStyle: 'solid', textAlign: 'center', cursor: 'default', transition: 'all 0.1s', background: 'rgba(0,0,0,.1)', boxShadow: '-1px -1px 3px rgba(0,0,0,.1) inset,1px 1px 3px rgba(255,255,255,.7) inset', top: '50%', bottom: 2, borderRadius: '0 0 2px 2px', borderWidth: '0 1px 1px 1px' }}>
                  <i style={{ position: 'absolute', top: '50%', left: '50%', width: 0, height: 0, borderWidth: '0.6ex 0.6ex 0 0.6ex', borderColor: 'rgba(0, 0, 0, 0.7) transparent transparent', borderStyle: 'solid', margin: '-0.3ex 0 0 -0.56ex' }} />
                </b>
              </span>
            </NumericInput>
          </div>
        </Num>
      </AutoField>
    </section>
  </OptGroup>
</OptGroupField>"
`;
